### PR TITLE
Closes multiple issues targeted to 0.11

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: docs/User/Project/Upgrading-0.8-to-0.9.md [hunk 1]
```diff
@@ -281,13 +281,17 @@ parser from the `Streamly.Data.Parser` module.
...
+>>> groupsBy eq fld = Stream.parseMany (Parser.groupBy (flip eq) fld)
 >>> groups = groupsBy (==)
 >>> groupsByRolling = Streamly.Internal.Data.Stream.groupsRollingBy
 >>> wordsBy p f = Stream.parseMany (Parser.wordBy p f)
 >>> chunksOf n f = Stream.foldMany (Fold.take n f)
 ```
 
+**Caution**: In `streamly-0.8`, The `groupsBy` comparison function took the new
+element as the first argument and the previous element as the second.  The order
+is flipped in the newer versions if streamly.
```
Comment (adithyaov) [inline docs/User/Project/Upgrading-0.8-to-0.9.md:293]: if -> of

File: core/src/Streamly/Internal/Data/Parser.hs [hunk 2]
```diff
@@ -1714,7 +1714,7 @@ data WordQuotedState s b a =
 {-# INLINE wordWithQuotes #-}
 wordWithQuotes :: (Monad m, Eq a) =>
        Bool            -- ^ Retain the quotes and escape chars in the output
-    -> (a -> a -> Maybe a)  -- ^ quote char -> escaped char -> translated char
+    -> (a -> Maybe a -> Maybe a)  -- ^ quote char -> escaped char -> translated char
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser.hs:1717]: Add a docspec test.

File: core/src/Streamly/Internal/Data/Parser.hs [hunk 3]
```diff
@@ -1867,7 +1886,19 @@ wordProcessQuotes :: (Monad m, Eq a) =>
     -> Parser a m b
 wordProcessQuotes =
     -- Escape the quote char itself
-    wordWithQuotes False (\q x -> if q == x then Just x else Nothing)
+    wordWithQuotes False escTransation
+
+    where
+
+    escTransation q mx =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser.hs:1893]: escTranslation**

File: core/src/Streamly/Internal/Data/Parser.hs [hunk 4]
```diff
@@ -1826,6 +1826,14 @@ wordWithQuotes keepQuotes tr escChar toRight isSep
         if n == 0
         then fmap (Done 0) $ ffinal s
         else err "wordWithQuotes: missing frame end"
+    extract (WordQuotedEsc s 0 ql _) =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser.hs:1829]: Should work for all `n` not only `0`

File: benchmark/Streamly/Benchmark/Data/Stream/Expand.hs [hunk 5]
```diff
@@ -238,6 +238,8 @@ o_1_space_applicative value =
         , benchIO "(<*) (sqrt n x sqrt n)" $ apDiscardSnd value
         , benchIO "(<*>) (sqrt n x sqrt n)" $ toNullAp value
         , benchIO "liftA2 (sqrt n x sqrt n)" $ apLiftA2 value
+        -- XXX Do we need these benchmarks?
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Stream/Expand.hs:241]: Yes.

File: benchmark/Streamly/Benchmark/Data/Stream/Transform.hs [hunk 6]
```diff
@@ -538,9 +548,13 @@ o_1_space_inserting value =
         , benchIOSink value "insertBy" (insertBy value 1)
         , benchIOSink value "interposeSuffix" (interposeSuffix value 1)
         , benchIOSink value "intercalateSuffix" (intercalateSuffix value 1)
+        , benchIOSinkPureSrc value "interspersePure" (intersperse value 1)
         ]
     ]
 
+-- XXX Shouldn't we use value16 here?
+-- XXX Initially classified as o_n_stack
+-- where value16 = round (fromIntegral value**(1/16::Double))
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Stream/Transform.hs:557]: Can run the benchmarks and check with `--long`

File: core/src/Streamly/Internal/Data/Array/Type.hs [hunk 7]
```diff
@@ -1000,8 +1026,8 @@ fromPureStream x = unsafePerformIO $ fmap unsafeFreeze (MA.fromPureStream x)
 --
 -- Note that this should be evaluated strictly to ensure that we do not hold
 -- the reference to the pointer in a lazy thunk.
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Type.hs:1028]: We can remove this comment now.

File: core/src/Streamly/Internal/Data/Array/Type.hs [hunk 8]
```diff
@@ -293,6 +297,30 @@ unsafeAsForeignPtr arr@Array{..} f =
             fptr = ForeignPtr addr# fptrContents
          in f fptr i
 
+mutableByteArrayContents# :: Exts.MutableByteArray# s -> Addr#
+{-# INLINE mutableByteArrayContents# #-}
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Type.hs:301]: put INLINE pragma before the signature.

File: core/src/Streamly/Internal/Data/Array.hs [hunk 9]
```diff
@@ -560,13 +560,10 @@ pinnedSerialize = encodeAs Pinned
...
-deserialize arr@(Array {..}) = unsafeInlineIO $ do
-    let lenArr = length arr
-    (off, val) <-
-        Serialize.deserializeAt arrStart arrContents (arrStart + lenArr)
-    assertM(off == arrStart + lenArr)
-    pure val
+deserialize :: Serialize a => Array Word8 -> (a, Array Word8)
+deserialize arr =
+    let (a, b) = unsafeInlineIO $ MA.deserialize (unsafeThaw arr)
+     in (a, unsafeFreeze b)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array.hs:566]: can just use `fmap unsafeFreeze`.

File: core/src/Streamly/Internal/Data/Array/Type.hs [hunk 10]
```diff
@@ -293,6 +297,30 @@ unsafeAsForeignPtr arr@Array{..} f =
...
+unsafeFromForeignPtr (ForeignPtr addr# _) i
+    | Ptr addr# == nullPtr || i == 0 = empty
+unsafeFromForeignPtr (ForeignPtr addr# (PlainPtr marr#)) len =
+    let off = I# (addr# `minusAddr#` mutableByteArrayContents# marr#)
+     in Array (Unboxed.MutByteArray marr#) off (off + len)
+unsafeFromForeignPtr (ForeignPtr addr# _) len =
+    unsafeInlineIO
+        $ fromStreamN len
+        $ D.mapM (peek . plusPtr (Ptr addr#))
+        $ D.fromList [0,1..(len - 1)]
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Type.hs:322]: We should use `memcpy` for better performance.

File: core/src/Streamly/Internal/Data/RingArray/Generic.hs [hunk 11]
```diff
@@ -37,7 +37,7 @@ import qualified Streamly.Internal.Data.Fold.Type as Fold
 import qualified Streamly.Internal.Data.MutArray.Generic as MutArray
 
 -- XXX Use MutableArray rather than keeping a MutArray here.
-data Ring a = Ring
+data RingArray a = RingArray
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/RingArray/Generic.hs:40]: Provide a type synonym for old users, if any?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Array.hs (hunk 9 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Array.hs:566]: I don't like using `fmap` on tuples. To me, tuple as a functor is not very intuitive.

File: core/src/Streamly/Internal/Data/RingArray.hs [hunk 12]
```diff
@@ -166,6 +167,9 @@ data RingArray a = RingArray
     , ringHead :: {-# UNPACK #-} !Int -- byte index in the array
     }
 
+{-# WARNING Ring "Please use RingArray instead." #-}
+type Ring = RingArray
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/RingArray.hs:171]: why not use DEPRECATED?

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**